Framework/Geometry/inc/MantidGeometry/Instrument/InfoIteratorBase.h

```diff
@@ -0,0 +1,86 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#ifndef MANTID_GEOMETRY_INFOITERATORBASE_H_
+#define MANTID_GEOMETRY_INFOITERATORBASE_H_
+
+#include <algorithm>
+#include <boost/iterator/iterator_facade.hpp>
+
+namespace Mantid {
+namespace Geometry {
+
+/** InfoIterator
+
+Base to allow users of the Info objects (DetectorInfo etc) access to data
+via a random access iterator.
+*/
+template <typename T, template <typename> class InfoItem>
+class InfoIteratorBase
```

Comment:
I have been wondering whether we could do this much simpler: Recently I discovered `boost::transform_iterator` (see https://github.com/mantidproject/dataset/blob/master/src/dataset.h#L84). If we combine this with `boost::counting_iterator`, can we eliminate the need for a custom iterator?

```cpp
auto DetectorInfo::begin() {
  return boost::make_transform_iterator(
      boost::counting_iterator<size_t>(0),
      makeDetectorInfoItem));
}
```

Reply:
I am not sure how `makeDetectorInfoItem` would get the instance though?

Reply:
Can we create the lambda `makeDetectorInfoItem` using a member function? Then we can capture `this` there.